### PR TITLE
Add Baseten Inference as Proxy

### PR DESCRIPTION
**Pull request by AlexKer**
## What problem(s) was I solving?

Add Baseten support. In response to: https://github.com/humanlayer/humanlayer/issues/486

## What user-facing changes did I ship?

Added Baseten, extending https://github.com/humanlayer/humanlayer/pull/453
- **Frontend**: Added Baseten provider option with model input and API key management
- **Backend**: Proxy requests through HumanLayer daemon to Baseten's OpenAI-compatible API
- **Authentication**: Uses Bearer token authentication with user's Baseten API key

## How I implemented it

Exactly the same as OpenRouter implementation. 
The main challenge was getting Claude CLI to route requests through the proxy. **Claude CLI requires BOTH environment variables to trigger proxy mode:**
- `ANTHROPIC_BASE_URL` → Points to our proxy endpoint  
- `ANTHROPIC_API_KEY` → Dummy value to activate proxy (so I added line 217 in `manager.go`)
Without both variables set, Claude CLI ignores the custom base URL and hits Anthropic directly.

## How to verify it

- [x] I have ensured `make check test` passes
- DeepSeek V3.1 model via Baseten
- FastAPI project file analysis and test generation
- 21K+ tokens processed successfully
- All linting/formatting/tests pass

## Description for the changelog

Added Baseten proxy support for model customization - Users can now use Baseten as an alternative LLM provider, support OSS models like DeepSeek v3.1, Kimi K2, Qwen3 Coder. 

## A picture of a cute animal (not mandatory but encouraged)
61QiUmahC6L

----

> [!IMPORTANT]
> Add Baseten proxy support for model customization, including API key management and request routing.
> 
>   - **Backend**:
>     - Add Baseten proxy support in `proxy.go` and `proxy_transform.go`.
>     - Update `manager.go` to handle Baseten API key and proxy configuration.
>   - **Frontend**:
>     - Add Baseten provider option in `CommandInput.tsx`.
>     - Manage Baseten API key in `useSessionLauncher.ts`.
>   - **Configuration**:
>     - Update `config.go` to check Baseten API key status.
>     - Modify `http-client.ts` to pass Baseten proxy configuration.
>   - **Misc**:
>     - Update `types.ts` to include Baseten in session parameters.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=humanlayer%2Fhumanlayer&utm_source=github&utm_medium=referral)<sup> for e34214cca7598a514639d0757ccbd3af746cec94. You can [customize](https://app.ellipsis.dev/humanlayer/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: hld/api/handlers/proxy.go [hunk 1]
```diff
@@ -198,6 +207,18 @@ func (h *ProxyHandler) handleNonStreamingProxy(c *gin.Context, sessionID string,
 			return
 		}
 		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", apiKey))
+	} else if strings.Contains(url, "inference.baseten.co") {
```
[the remark this answers is not in the document]
Comment (AlexKer) [reply hld/api/handlers/proxy.go]: refactored!

File: hld/session/manager.go
Comment (samdickson22) [inline hld/session/manager.go:1]: todo: verify this

File: hld/session/manager.go [hunk 2]
```diff
@@ -1494,6 +1496,8 @@ func (m *Manager) ContinueSession(ctx context.Context, req ContinueSessionConfig
 		}
 		proxyURL := fmt.Sprintf("http://localhost:%d/api/v1/anthropic_proxy/%s", httpPort, sessionID)
 		config.Env["ANTHROPIC_BASE_URL"] = proxyURL
```
Comment (samdickson22) [inline hld/session/manager.go:1498]: todo: verify this

File: hld/session/manager.go [hunk 3]
```diff
@@ -213,6 +213,8 @@ func (m *Manager) LaunchSession(ctx context.Context, config LaunchSessionConfig)
 		}
 		proxyURL := fmt.Sprintf("http://localhost:%d/api/v1/anthropic_proxy/%s", httpPort, sessionID)
 		claudeConfig.Env["ANTHROPIC_BASE_URL"] = proxyURL
```
Comment (samdickson22) [inline hld/session/manager.go:215]: todo: verify this

File: humanlayer-wui/src/components/CommandInput.tsx [hunk 4]
```diff
@@ -45,6 +46,8 @@ export default function CommandInput({
   const { paths: recentPaths } = useRecentPaths()
   const [showApiKey, setShowApiKey] = useState(false)
```
Comment (samdickson22) [inline humanlayer-wui/src/components/CommandInput.tsx]: rename general names to openrouter if its referring to that, if its actually general functions, move baseten stuff inside them

File: humanlayer-wui/src/components/CommandInput.tsx [hunk 5]
```diff
@@ -178,6 +182,15 @@ export default function CommandInput({
                 placeholder="e.g., openai/gpt-oss-120b"
                 disabled={isLoading}
               />
```
Comment (samdickson22) [inline humanlayer-wui/src/components/CommandInput.tsx:189]: maybe create shared model selector component for openrouter and baseten

File: hld/session/manager.go [hunk 6]
```diff
@@ -213,6 +213,8 @@ func (m *Manager) LaunchSession(ctx context.Context, config LaunchSessionConfig)
...
+		// Claude CLI needs an API key to trigger requests, even though proxy handles auth, set dummy key
```
Comment (samdickson22) [inline hld/session/manager.go:216]: are you sure? i feel as though if this was the case then it would have failed before using openrouter

File: hld/session/manager.go [hunk 7]
```diff
@@ -1494,6 +1496,8 @@ func (m *Manager) ContinueSession(ctx context.Context, req ContinueSessionConfig
 		}
 		proxyURL := fmt.Sprintf("http://localhost:%d/api/v1/anthropic_proxy/%s", httpPort, sessionID)
 		config.Env["ANTHROPIC_BASE_URL"] = proxyURL
+		// Claude CLI needs an API key to trigger requests, even though proxy handles auth
```
Comment (samdickson22) [inline hld/session/manager.go:1499]: are you sure? i feel as though if this was the case then it would have failed before using openrouter

File: hld/session/manager.go (hunk 6 above)
Comment (AlexKer) [reply hld/session/manager.go:216]: it was failing for me with using openrouter

Comment (AlexKer) [reply hld/session/manager.go:216]: I don't have ANTHROPIC_API_KEY set, if you have a set, maybe it would have worked. See above: https://github.com/humanlayer/humanlayer/pull/501#issue-3353375796 for explanation.

File: humanlayer-wui/src/components/CommandInput.tsx (hunk 5 above)
Comment (AlexKer) [reply humanlayer-wui/src/components/CommandInput.tsx:189]: Done

File: humanlayer-wui/src/components/CommandInput.tsx (hunk 4 above)
Comment (AlexKer) [reply humanlayer-wui/src/components/CommandInput.tsx]: don't really know what this means re: those 2 lines

Comment (AlexKer) [reply humanlayer-wui/src/components/CommandInput.tsx]: don't know what this means? could you fix?

**Comment by AlexKer** [comment]
@samdickson22 I ran `make check test` and got this error. I think it's coming from main. I've fixed it in the latest commit so all the tests are passing.
src/components/internal/SessionDetail/SessionDetail.tsx(228,53): error TS2353: Object literal may only specify known properties, and 'additionalDirectories' does not exist in type 'Partial<Session>'.
src/components/internal/SessionDetail/SessionDetail.tsx(982,38): error TS2339: Property 'additionalDirectories' does not exist on type 'Session'.
src/components/internal/SessionDetail/SessionDetail.tsx(1073,38): error TS2339: Property 'additionalDirectories' does not exist on type 'Session'.

File: hld/session/manager.go (hunk 6 above)
Comment (samdickson22) [reply hld/session/manager.go:216]: were you signed into claude code

Comment (AlexKer) [reply hld/session/manager.go:216]: don't think so

**Comment by samdickson22** [comment]
> @samdickson22 I ran `make check test` and got this error. I think it's coming from main. I've fixed it in the latest commit so all the tests are passing. src/components/internal/SessionDetail/SessionDetail.tsx(228,53): error TS2353: Object literal may only specify known properties, and 'additionalDirectories' does not exist in type 'Partial'. src/components/internal/SessionDetail/SessionDetail.tsx(982,38): error TS2339: Property 'additionalDirectories' does not exist on type 'Session'. src/components/internal/SessionDetail/SessionDetail.tsx(1073,38): error TS2339: Property 'additionalDirectories' does not exist on type 'Session'.




run "make codelayer-dev" again, it will run database migrations

File: hld/session/manager.go (hunk 6 above)
Comment (AlexKer) [reply hld/session/manager.go:216]: you can reject this line if you want, as long as it's working on your end, that's fine

Comment (samdickson22) [reply hld/session/manager.go:216]: no, its probably a good thing to have - never considered people running codelayer not logged into claude code since we haven't had other providers

Comment (AlexKer) [reply hld/session/manager.go:216]: assuming that folks just want to use coderlayer without claude code, then I agree it should be compatible. kind of an edge case

**Comment by AlexKer** [comment]
> > @samdickson22 I ran `make check test` and got this error. I think it's coming from main. I've fixed it in the latest commit so all the tests are passing. src/components/internal/SessionDetail/SessionDetail.tsx(228,53): error TS2353: Object literal may only specify known properties, and 'additionalDirectories' does not exist in type 'Partial'. src/components/internal/SessionDetail/SessionDetail.tsx(982,38): error TS2339: Property 'additionalDirectories' does not exist on type 'Session'. src/components/internal/SessionDetail/SessionDetail.tsx(1073,38): error TS2339: Property 'additionalDirectories' does not exist on type 'Session'.
> 
> run "make codelayer-dev" again, it will run database migrations

ah I see. yep it went away. the latest commit remains the same

**Comment by AlexKer** [comment]
> > @samdickson22 I ran `make check test` and got this error. I think it's coming from main. I've fixed it in the latest commit so all the tests are passing. src/components/internal/SessionDetail/SessionDetail.tsx(228,53): error TS2353: Object literal may only specify known properties, and 'additionalDirectories' does not exist in type 'Partial'. src/components/internal/SessionDetail/SessionDetail.tsx(982,38): error TS2339: Property 'additionalDirectories' does not exist on type 'Session'. src/components/internal/SessionDetail/SessionDetail.tsx(1073,38): error TS2339: Property 'additionalDirectories' does not exist on type 'Session'.
> 
> run "make codelayer-dev" again, it will run database migrations

ah I see. yep it went away, thanks! My latest commit remains the same, so it's ready to go.

**Comment by samdickson22** [comment]
> > > @samdickson22 I ran `make check test` and got this error. I think it's coming from main. I've fixed it in the latest commit so all the tests are passing. src/components/internal/SessionDetail/SessionDetail.tsx(228,53): error TS2353: Object literal may only specify known properties, and 'additionalDirectories' does not exist in type 'Partial'. src/components/internal/SessionDetail/SessionDetail.tsx(982,38): error TS2339: Property 'additionalDirectories' does not exist on type 'Session'. src/components/internal/SessionDetail/SessionDetail.tsx(1073,38): error TS2339: Property 'additionalDirectories' does not exist on type 'Session'.
> > 
> > 
> > run "make codelayer-dev" again, it will run database migrations
> 
> ah I see. yep it went away, thanks! My latest commit remains the same, so it's ready to go.

did it regenerate any files?

**Comment by AlexKer** [comment]
> > > > @samdickson22 I ran `make check test` and got this error. I think it's coming from main. I've fixed it in the latest commit so all the tests are passing. src/components/internal/SessionDetail/SessionDetail.tsx(228,53): error TS2353: Object literal may only specify known properties, and 'additionalDirectories' does not exist in type 'Partial'. src/components/internal/SessionDetail/SessionDetail.tsx(982,38): error TS2339: Property 'additionalDirectories' does not exist on type 'Session'. src/components/internal/SessionDetail/SessionDetail.tsx(1073,38): error TS2339: Property 'additionalDirectories' does not exist on type 'Session'.
> > > 
> > > 
> > > run "make codelayer-dev" again, it will run database migrations
> > 
> > 
> > ah I see. yep it went away, thanks! My latest commit remains the same, so it's ready to go.
> 
> did it regenerate any files?

no I regenerated all the files in this commit [5338c0d](https://github.com/humanlayer/humanlayer/pull/501/commits/5338c0d25eef0cc3e55601d5d3180f6858b157bd). I ran `make codelayer-dev` then `make check test`, nothing changed.

**Comment by AlexKer** [comment]
> > > > @samdickson22 I ran `make check test` and got this error. I think it's coming from main. I've fixed it in the latest commit so all the tests are passing. src/components/internal/SessionDetail/SessionDetail.tsx(228,53): error TS2353: Object literal may only specify known properties, and 'additionalDirectories' does not exist in type 'Partial'. src/components/internal/SessionDetail/SessionDetail.tsx(982,38): error TS2339: Property 'additionalDirectories' does not exist on type 'Session'. src/components/internal/SessionDetail/SessionDetail.tsx(1073,38): error TS2339: Property 'additionalDirectories' does not exist on type 'Session'.
> > > 
> > > 
> > > run "make codelayer-dev" again, it will run database migrations
> > 
> > 
> > ah I see. yep it went away, thanks! My latest commit remains the same, so it's ready to go.
> 
> did it regenerate any files?

no I regenerated all the files in this commit [5338c0d](https://github.com/humanlayer/humanlayer/pull/501/commits/5338c0d25eef0cc3e55601d5d3180f6858b157bd). I just ran make codelayer-dev then make check test, nothing changed.

**Comment by samdickson22** [comment]
> > > > > @samdickson22 I ran `make check test` and got this error. I think it's coming from main. I've fixed it in the latest commit so all the tests are passing. src/components/internal/SessionDetail/SessionDetail.tsx(228,53): error TS2353: Object literal may only specify known properties, and 'additionalDirectories' does not exist in type 'Partial'. src/components/internal/SessionDetail/SessionDetail.tsx(982,38): error TS2339: Property 'additionalDirectories' does not exist on type 'Session'. src/components/internal/SessionDetail/SessionDetail.tsx(1073,38): error TS2339: Property 'additionalDirectories' does not exist on type 'Session'.
> > > > 
> > > > 
> > > > run "make codelayer-dev" again, it will run database migrations
> > > 
> > > 
> > > ah I see. yep it went away, thanks! My latest commit remains the same, so it's ready to go.
> > 
> > 
> > did it regenerate any files?
> 
> no I regenerated all the files in this commit [5338c0d](https://github.com/humanlayer/humanlayer/pull/501/commits/5338c0d25eef0cc3e55601d5d3180f6858b157bd). I just ran make codelayer-dev then make check test, nothing changed.

perfect can you do the CLA so checks will run

**Comment by AlexKer** [comment]
> > > > > > @samdickson22 I ran `make check test` and got this error. I think it's coming from main. I've fixed it in the latest commit so all the tests are passing. src/components/internal/SessionDetail/SessionDetail.tsx(228,53): error TS2353: Object literal may only specify known properties, and 'additionalDirectories' does not exist in type 'Partial'. src/components/internal/SessionDetail/SessionDetail.tsx(982,38): error TS2339: Property 'additionalDirectories' does not exist on type 'Session'. src/components/internal/SessionDetail/SessionDetail.tsx(1073,38): error TS2339: Property 'additionalDirectories' does not exist on type 'Session'.
> > > > > 
> > > > > 
> > > > > run "make codelayer-dev" again, it will run database migrations
> > > > 
> > > > 
> > > > ah I see. yep it went away, thanks! My latest commit remains the same, so it's ready to go.
> > > 
> > > 
> > > did it regenerate any files?
> > 
> > 
> > no I regenerated all the files in this commit [5338c0d](https://github.com/humanlayer/humanlayer/pull/501/commits/5338c0d25eef0cc3e55601d5d3180f6858b157bd). I just ran make codelayer-dev then make check test, nothing changed.
> 
> perfect can you do the CLA so checks will run

I did it like 3 times, not sure why the server is not registering. even tried on different browsers. let me try again now

**Comment by AlexKer** [comment]
@samdickson22 I've already signed, but it's not showing up here. 
Screenshot 2025-08-27 at 10 29 08 PM
Document already shows "agreed" here:
Screenshot 2025-08-27 at 10 25 48 PM

**Comment by samdickson22** [comment]
> @samdickson22 I've already signed, but it's not showing up here. Screenshot 2025-08-27 at 10 29 08 PM Document already shows "agreed" here: Screenshot 2025-08-27 at 10 25 48 PM

the commit came from the email alexker@mac.mynetworksettings.com, do you have it added to your GH?

**Comment by AlexKer** [comment]
@samdickson22 [alexker@mac.mynetworksettings.com](mailto:alexker@mac.mynetworksettings.com) is the default email, I didn't have my github username and email set locally, so that's why the commit was coming from that email and Alex Ker instead of AlexKer. Not sure how to fix this since I've committed already. I don't have access to alexker@mac.mynetworksettings.com, so I can't add it to my account

**Comment by samdickson22** [comment]
> @samdickson22 [alexker@mac.mynetworksettings.com](mailto:alexker@mac.mynetworksettings.com) is the default email, I didn't have my github username and email set locally, so that's why the commit was coming from that email and Alex Ker instead of AlexKer. Not sure how to fix this since I've committed already. I don't have access to alexker@mac.mynetworksettings.com, so I can't add it to my account


Check who ur logged in as by running:

git config user.name
And
git config user.email


Run

git config user.email "new-email@example.com"

If it's wrong

Then

git commit --amend --author="Your Name <new-email@example.com>"

If it's your most recent commit

**Comment by samdickson22** [comment]
try "gh auth login" - i see its back to your mac email

**Comment by AlexKer** [comment]
@samdickson22 I tried to revise the history with git commit amend and forced push, CLA registered it, it looks fine now

**Comment by samdickson22** [comment]
> @samdickson22 I tried to revise the history with git commit amend and forced push, CLA registered it, it looks fine now

it was fine for that one commit, its still showing unsigned for me

**Comment by samdickson22** [comment]
> > @samdickson22 I tried to revise the history with git commit amend and forced push, CLA registered it, it looks fine now
> 
> it was fine for that one commit, its still showing unsigned for me

Screenshot 2025-08-27 at 8 09 26 PM

**Comment by AlexKer** [comment]
@samdickson22 that was painful but think we're good now!

**Comment by samdickson22** [comment]
Screenshot 2025-08-27 at 8 37 58 PM


can you add baseten to this model / provider selector screen? other than that it seems to work when i make a session starting with it

**Comment by AlexKer** [comment]
@samdickson22 thank you for the catch, should be fixed now. just tested locally

File: humanlayer-wui/src/components/CommandInput.tsx (hunk 4 above)
Comment (dexhorthy) [reply humanlayer-wui/src/components/CommandInput.tsx]: i'll add a Todo here

**Comment by dexhorthy** [comment]
other notes:
- fixing conflicts
- update copy on advanced routers to warn that its experimental
- adding modal warning "using non-claude models is experimental"

will load this in and get it merged
